add ability to push a mode if the current mode is an ancestor

samdow opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79143
* #78847
* __->__ #78822

Discussed briefly [here](https://github.com/pytorch/pytorch/pull/78677#discussion_r887431040) this lets a user set a mode whose inner is already set if that inner is the current mode or its an ancestor. This will be necessary for some more complicated use cases of composition and nice for non-lexical scoping

Commit:
add ability to push a mode if the current mode is an ancestor

[ghstack-poisoned]

Commit:
Update on "add ability to push a mode if the current mode is an ancestor"




[ghstack-poisoned]

Commit:
Update on "add ability to push a mode if the current mode is an ancestor"




[ghstack-poisoned]

Commit:
Update on "add ability to push a mode if the current mode is an ancestor"




[ghstack-poisoned]

Commit:
Update on "add ability to push a mode if the current mode is an ancestor"


Discussed briefly [here](https://github.com/pytorch/pytorch/pull/78677#discussion_r887431040) this lets a user set a mode whose inner is already set if that inner is the current mode or its an ancestor. This will be necessary for some more complicated use cases of composition and nice for non-lexical scoping

[ghstack-poisoned]

Commit:
Update on "add ability to push a mode if the current mode is an ancestor"


Discussed briefly [here](https://github.com/pytorch/pytorch/pull/78677#discussion_r887431040) this lets a user set a mode whose inner is already set if that inner is the current mode or its an ancestor. This will be necessary for some more complicated use cases of composition and nice for non-lexical scoping

[ghstack-poisoned]

ezyang commented on test/test_python_dispatch.py:
```diff
@@ -1285,6 +1301,7 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
                 called += 1
                 return super().__torch_dispatch__(func, types, args, kwargs)
 
+        print(torch._C._get_torch_dispatch_mode())
```
kill me

ezyang reviewed (approved):
One thought: should we always allow pushing in this way, or should you explicitly have to say something like `mode.restore()` to trigger this.

ezyang commented:
one reason I think it might be reasonable to have a separate API is that when you saving mode to use it later it's a pretty different use case than pushing the mode. We could easily explain that you can't reuse a mode and if you want to restore it you should use the restore method in the error message.

zou3519 commented on torch/utils/_python_dispatch.py:
```diff
@@ -143,20 +143,29 @@ class TorchDispatchMode(metaclass=TorchDispatchModeMeta):
-        self.inner = old
+        if hasattr(self, "inner"):
+            assert hasattr(self, "ancestors")
+            if old is not None and old not in self.ancestors:
+                raise RuntimeError(f"{self} has already been used as a mode and is not valid in the current state, " +
+                                   "because the current mode is not its ancestor. Please use a fresh version")
```
We've got these new error messages, but I don't see a test testing them. Should we add some?

Commit:
Update on "add ability to push a mode if the current mode is an ancestor"


Discussed briefly [here](https://github.com/pytorch/pytorch/pull/78677#discussion_r887431040) this lets a user set a mode whose inner is already set if that inner is the current mode or its an ancestor. This will be necessary for some more complicated use cases of composition and nice for non-lexical scoping

[ghstack-poisoned]

Commit:
Update on "add ability to push a mode if the current mode is an ancestor"


Discussed briefly [here](https://github.com/pytorch/pytorch/pull/78677#discussion_r887431040) this lets a user set a mode whose inner is already set if that inner is the current mode or its an ancestor. This will be necessary for some more complicated use cases of composition and nice for non-lexical scoping

[ghstack-poisoned]

samdow commented:
Moved to `restore` so now a user will have to do

`with x.restore()` or `with restore(x):` if `x` has already been used

Commit:
Update on "add ability to push a mode if the current mode is an ancestor"


Discussed briefly [here](https://github.com/pytorch/pytorch/pull/78677#discussion_r887431040) this lets a user set a mode whose inner is already set if that inner is the current mode or its an ancestor. This will be necessary for some more complicated use cases of composition and nice for non-lexical scoping

[ghstack-poisoned]

Commit:
Update on "add ability to push a mode if the current mode is an ancestor"


Discussed briefly [here](https://github.com/pytorch/pytorch/pull/78677#discussion_r887431040) this lets a user set a mode whose inner is already set if that inner is the current mode or its an ancestor. This will be necessary for some more complicated use cases of composition and nice for non-lexical scoping

[ghstack-poisoned]

Commit:
Update on "add ability to push a mode if the current mode is an ancestor"


Discussed briefly [here](https://github.com/pytorch/pytorch/pull/78677#discussion_r887431040) this lets a user set a mode whose inner is already set if that inner is the current mode or its an ancestor. This will be necessary for some more complicated use cases of composition and nice for non-lexical scoping

[ghstack-poisoned]

Commit:
Update on "add ability to push a mode if the current mode is an ancestor"


Discussed briefly [here](https://github.com/pytorch/pytorch/pull/78677#discussion_r887431040) this lets a user set a mode whose inner is already set if that inner is the current mode or its an ancestor. This will be necessary for some more complicated use cases of composition and nice for non-lexical scoping

[ghstack-poisoned]

samdow commented:
pytorchbot merge -a

Commit:
Update on "add ability to push a mode if the current mode is an ancestor"


Discussed briefly [here](https://github.com/pytorch/pytorch/pull/78677#discussion_r887431040) this lets a user set a mode whose inner is already set if that inner is the current mode or its an ancestor. This will be necessary for some more complicated use cases of composition and nice for non-lexical scoping

[ghstack-poisoned]

samdow commented:
pytorchbot merge

samdow commented:
@pytorchbot merge
